design/multi-factor-admin-and-step-up.md

```diff
@@ -0,0 +1,582 @@
+within the admin console (what this implies for flows and clients), and the
+implications for users.
+
+### Step up mechanisms
+
+Step-up is the ability to allow access to clients or resources based on an authentication
+level of a user. It can be described by two example use cases:
+1. A user has a session in Keycloak for a *client A* after having logged in with username and password. However, when connecting to a *client B* which demands greater security, the user is asked for a second authentication factor before proceeding. Had the user connected initially to *client B*, he would have had to provide username/password **and** the second factor.
+1. A user has a session in Keycloak with a *client C*, that was logged in with username and password. However, when doing an operation with more risk *client C* asks Keycloak to provide a second factor authentication for the user. Note that the client may also ask for a full 2 factor authentication: username/password + 2nd factor. A good example for this case would be a bank that allows a user to browse his transaction history with only username/password, but which requires a 2nd factor to access the payment functionality.
+
+In many ways, step-up can be viewed as an authorization which is enabled by stronger
+authentication to be granted. Use case 1 can be handled only through internal configurations
+within Keycloak but the use case 2 requires support from both Keycloak and the client.
+Conceptually, the mechanisms to support both use cases must be present within Keycloak.
+
+Use case 2 is supported by both the SAML and OIDC protocols. Within these protocols
+there are three concepts that are represented that enable step-up:
+* Level of Authentication (LoA)
+* Specifying authentication categories
+* Forcing authentication
+
+#### Level of Authentication
+
+This defines an agreed upon level of authentication (also called level of assurance)
+between an IDP and client. There is no single formal document that specifies the
+requirements for an authentication.
+For example [NIST's SP 800-63-3](https://pages.nist.gov/800-63-3/) has three different
+categories of three levels each, while
+[ISO/IEC 29115](https://www.iso.org/standard/45138.html) defines 4 assurance levels.
+The basic idea is the manner in which an authentication is done or which authenticators
+are used causes lower or greater assurances about the identity of the user. This level
+can then be used to determine if a user can or not perform a requested action.
+
+OIDC supports LoA with the optional Authentication Context Class `acr` claim,
+or with the acr_values parameter as defined in the
+[core specification](https://openid.net/specs/openid-connect-core-1_0.html#acrSemantics).
+
+SAML supports LoA with the [Identity Assurance Profiles](http://docs.oasis-open.org/security/saml/Post2.0/sstc-saml-assurance-profile.html) part of the protocol, and with the
+`<RequestedAuthnContext>` element.
+
+#### Specifying authenticator categories
```

Comment:
Regarding OIDC, I think it should be enough to use `acr_values` as the mean to enforce a specific authentication method, and as a consequence a security level.

The `amr` claim should hold values for each authentication method used to assert the user's identity.

Reply:
The protocol specifies both the acr_values and requesting the acr claim as an essential claim as valid methods for requesting a security level, so why not implement both? Aside from the fact that it is more work, obviously. It would allow for greater compatibility with whatever SPs may wish to use this feature.

For the amr you are correct, I thought I described it thus in the document, but if you think that it's not clear enough I'll rework the text to make it clearer.

Reply:
Yeah, we should do both. Where `acr_values` is what we would use to request tokens with a specific `acr` value.